### ISSUE TITLE
Questions about OOD Detector?

### ISSUE THREAD

**Issue by Gaop970222** [zzwjames/DPGBA#3]
Hi,

Thank you so much for sharing your code — it's well-structured and has been very helpful to my work. While using it, I came across a couple of points that I’m unsure about, and I’d really appreciate your clarification:

1. In the MLP-based OOD detector, I noticed that the final `nn.ReLU()` and `nn.Linear(nhid, nclass)` layers were removed. I’m curious about the reasoning behind this design choice — is it intentional for training stability or some other purpose?

2. I also noticed a possible inconsistency between the inner and outer loops. In the inner loop, the trigger features are labeled as 0:

```
ood_x = torch.cat([selected_features,trojan_feat]).detach()
output_detector = self.ood_detector(ood_x)
ood_labels = torch.cat([torch.ones(len(trojan_feat), device=self.device), torch.zeros(len(trojan_feat), device=self.device)])
```

But in the outer loop, the same trojan_feat seems to be assigned label 1:

```
update_feat = torch.cat([features,trojan_feat])....
output_detector= self.ood_detector(update_feat)...
loss_dis = F.nll_loss(output_detector[-len(trojan_feat):], torch.ones(len(trojan_feat), device=self.device).long())
```

I'm not sure if this is a deliberate technique (e.g., for contrastive training), or if I misunderstood something. I’d be grateful if you could kindly shed some light on this.

Thanks again for your time and for making your work available — I really appreciate it!

**Comment by Gaop970222** [comment]
Hi Zhang：

I got the reason of the question 2, but still have problem with Q1. Thank you very much!

Best Regards

**Comment by zzwjames** [comment]
Hi,

Thanks for your interest in our work! For Q1, there is no specific reason behind that design, I believe that part would not impact the result a lot. Thanks!

**Comment by Gaop970222** [comment]
Thank you very much!
